asreview/json_logging.py

```diff
@@ -0,0 +1,238 @@
+            "train_idx": train_idx.tolist(),
+            "proba": proba.tolist(),
+        }
+        self._add_to_log(new_dict, query_i)
+
+    def n_queries(self):
+        """Get the number of stored queries."""
+        return len(self._log_dict["results"])
+
+    def review_state(self):
+        """Get the state of the review variables from the internal state."""
+        labels = np.array(self._log_dict["labels"], dtype=np.int)
+
+        train_idx = []
+        query_src = {}
+        for query_i, res in enumerate(self._log_dict["results"]):
+            if "labelled" not in res:
+                continue
+            label_idx = [x[0] for x in res["labelled"]]
+            inclusions = [x[1] for x in res["labelled"]]
+            label_meth = [x[2] for x in res["labelled"]]
+            for i, meth in enumerate(label_meth):
+                if meth not in query_src:
+                    query_src[meth] = []
+                query_src[meth].append(label_idx[i])
+                labels[label_idx[i]] = inclusions[i]
+            train_idx.extend(label_idx)
+        if query_i > 0 and "labelled" not in self._log_dict["results"][-1]:
+            query_i -= 1
+
+        train_idx = np.array(train_idx, dtype=np.int)
+        return labels, train_idx, query_src, query_i
+
+    def get(self, variable, query_i=None, idx=None):
+        """Get data from the internal state."""
+        if query_i is not None:
+            res = self._log_dict["results"][query_i]
+
+        array = None
+        if variable == "label_methods":
+            array = res["labelled"]
+            array = np.array([x[2] for x in array], dtype=str)
+        elif variable in ["label_idx", "labelled"]:
+            array = res["labelled"]
+            array = np.array([x[0] for x in array], dtype=np.int)
+        elif variable == "labels":
+            array = np.array(self._log_dict["labels"], dtype=np.int)
+        elif variable == "final_labels":
+            array = np.array(self._log_dict["final_labels"], dtype=np.int)
+        elif variable == "proba":
+            array = np.array(res["proba"], dtype=np.float)
+        elif variable == "train_idx":
+            array = np.array(res["train_idx"], dtype=np.int)
+        elif variable == "pool_idx":
+            array = np.array(res["pool_idx"], dtype=np.int)
+        if array is None:
+            return None
+        if idx is not None:
+            return array[idx]
+        return array
+
+    def create_structure(self):
+        self._log_dict = OrderedDict({
+            "time": {"start_time": str(datetime.now())},
+            "version": self.version,
+            "software_version": asreview.__version__,
+            "settings": {},
+            "results": [],
+        })
+
+    def close(self):
+        if not self.read_only:
+            self.save()
+
+    def save(self):
+        print(self)
+
+
+class JSON_Logger(Dict_Logger):
+    def __init__(self, log_fp, *_, read_only=False, **__):
```

Comment:
I don't like this syntax :) Does `def __init__(self, log_fp, read_only=False):` fail?

Reply:
Not yet I think? If one adds options to the HDF5 one that are not applicable, then it would start to break.